File: ios/lib/ReactNativeCameraKit/CKCompressedImage.m
#import "CKCompressedImage.h"

@implementation CKCompressedImage

- (instancetype)initWithImage:(UIImage *)image imageQuality:(NSString*)imageQuality
{
    if(self = [super init]) {
        CGFloat max = 1200.0f;
        if ([imageQuality isEqualToString:@"high"]) {
            max = 1200.0f;
        }
        else if ([imageQuality isEqualToString:@"medium"]) {
            max = 800.0f;
        }
        else {
            _image = image;
            _data = UIImageJPEGRepresentation(image, 1.0f);
        }
        float actualHeight = image.size.height;
        float actualWidth = image.size.width;

        float imgRatio = actualWidth/actualHeight;

        float newHeight = (actualHeight > actualWidth) ? max : max/imgRatio;
        float newWidth = (actualHeight > actualWidth) ? max*imgRatio : max;


        CGRect rect = CGRectMake(0.0, 0.0, newWidth, newHeight);
        UIGraphicsBeginImageContext(rect.size);
        [image drawInRect:rect];
        _image = UIGraphicsGetImageFromCurrentImageContext();
        UIGraphicsEndImageContext();
        _data = UIImageJPEGRepresentation(_image, 0.85f);
    }

    return self;
}


@end

File: ios/ReactNativeCameraKit/CKCompressedImage.swift
//
//  CKCompressedImage.swift
//  ReactNativeCameraKit
//

import UIKit

enum ImageQuality: String {
    case high
    case medium
    case original

    init(from string: String) {
        self = ImageQuality(rawValue: string) ?? .original
    }
}

struct CKCompressedImage {
    let image: UIImage
    let data: Data?

    init(inputImage: UIImage, imageQuality: ImageQuality) {
        var max: CGFloat = 1200.0

        switch imageQuality {
        case .high:
            max = 1200.0
        case .medium:
            max = 800.0
        case .original:
            image = inputImage
            data = inputImage.jpegData(compressionQuality: 1.0)
            return
        }

        let actualHeight = inputImage.size.height
        let actualWidth = inputImage.size.width

        let imgRatio = actualWidth / actualHeight

        let newHeight = (actualHeight > actualWidth) ? max : max / imgRatio
        let newWidth = (actualHeight > actualWidth) ? max * imgRatio : max

        let rect = CGRect(x: 0.0, y: 0.0, width: newWidth, height: newHeight)
        UIGraphicsBeginImageContext(rect.size)
        inputImage.draw(in: rect)
        image = UIGraphicsGetImageFromCurrentImageContext()!
        UIGraphicsEndImageContext()
        self.data = image.jpegData(compressionQuality: 0.85)
    }
}
